source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp

```diff
@@ -0,0 +1,259 @@
+  std::advance(it_points, pos);
+  std::advance(it_times, pos);
+
+  this->points_.insert(it_points, new_point);
+
+  auto previous_time = this->times_[pos - 1];
+  this->times_.insert(it_times, previous_time + new_time);
+
+  for (unsigned int i = pos + 1; i <= this->points_.size(); i++) {
+    this->times_[i] += new_time;
+  }
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::delete_point() {
+  this->set_empty(false);
+  if (!this->points_.empty()) {
+    this->points_.pop_back();
+  }
+  if (!this->times_.empty()) {
+    this->times_.pop_back();
+  }
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::clear() {
+  this->points_.clear();
+  this->times_.clear();
+}
+
+template<typename TrajectoryT>
+inline const std::deque<TrajectoryT>& TrajectoryBase<TrajectoryT>::get_points() const {
+  return this->points_;
+}
+
+template<typename TrajectoryT>
+inline const TrajectoryT& TrajectoryBase<TrajectoryT>::get_point(unsigned int index) const {
+  return this->points_[index];
+}
+
+template<typename TrajectoryT>
+inline TrajectoryT& TrajectoryBase<TrajectoryT>::get_point(unsigned int index) {
+  return this->points_[index];
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline bool TrajectoryBase<TrajectoryT>::set_point(
+    unsigned int index, const TrajectoryT& point, const std::chrono::duration<int64_t, DurationT>& new_time) {
+  if (index < this->points_.size()) {
+    this->points_[index] = point;
+    if (index == 0) {
+      this->times_[index] = new_time;
+    } else {
+      this->times_[index] = this->times_[index - 1] + new_time;
+    }
+    for (unsigned int i = index + 1; i < this->points_.size(); ++i) {
+      this->times_[i] += this->times_[index - 1];
+    }
+    return true;
+  }
+  return false;
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline bool TrajectoryBase<TrajectoryT>::set_points(
+    const std::vector<TrajectoryT>& points, const std::vector<std::chrono::duration<int64_t, DurationT>>& new_times) {
+  if (points.size() != new_times.size()) {
+    return false;
+  }
+  this->clear();
+  for (unsigned int i = 0; i < points.size(); ++i) {
+    this->add_point(points[i], new_times[i]);
+  }
+  return true;
+}
+
+template<typename TrajectoryT>
+inline const std::deque<std::chrono::nanoseconds>& TrajectoryBase<TrajectoryT>::get_times() const {
+  return this->times_;
+}
+
+template<typename TrajectoryT>
+inline int TrajectoryBase<TrajectoryT>::get_size() const {
+  return this->points_.size();
+}
+
+template<typename TrajectoryT>
+inline const std::pair<TrajectoryT, std::chrono::nanoseconds>
+TrajectoryBase<TrajectoryT>::operator[](unsigned int idx) const {
+  return std::make_pair(this->points_[idx], this->times_[idx]);
+}
+
+template<typename TrajectoryT>
+inline std::pair<TrajectoryT, std::chrono::nanoseconds> TrajectoryBase<TrajectoryT>::operator[](unsigned int idx) {
+  this->set_empty(false);
```

Comment:
What is this for?

Reply:
that was legacy code, I assume it was for cases where `trajectory[n]` would have to return a copy of the data and not a reference.